### ISSUE TITLE
has_many through associations ignore the joins option of the through association

### ISSUE THREAD

**Issue by jeroenj** [rails/rails#33525]
### Steps to reproduce

```ruby
begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true

  create_table :users, force: true

  create_table :sources, force: true do |t|
    t.boolean :online, null: false, default: true
  end

  create_table :comments, force: true do |t|
    t.references :post
    t.references :user
    t.references :source
  end
end

class Post < ActiveRecord::Base
  has_many :comments
  has_many :online_comments, ->{ joins(:source).where(sources: {online: true}) }, class_name: "Comment"
  has_many :online_users, through: :online_comments, source: :user
end

class User < ActiveRecord::Base
end

class Source < ActiveRecord::Base
end

class Comment < ActiveRecord::Base
  belongs_to :source
  belongs_to :post
  belongs_to :user
end

class BugTest < Minitest::Test
  def test_association
    Comment.create!(post: Post.create!, source: Source.create!(online: true), user: User.create!)
    post = Post.first!
    assert_equal post.online_comments, [Comment.first!]
    assert_equal post.online_users, [User.first!] # Fails with `ActiveRecord::StatementInvalid: SQLite3::SQLException: no such column: sources.online: SELECT "users".* FROM "users" INNER JOIN "comments" ON "users"."id" = "comments"."user_id" WHERE "comments"."post_id" = ? AND "sources"."online" = ?`
  end
end
```
(Guidelines for creating a bug report are [available
here](http://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#creating-a-bug-report))

### Expected behavior

When using the `has_many :online_users` association on `Post` in the example above you'd expect the `joins(:source)` from the `online_comments` association to be there in the query (just like the conditionals set there).

### Actual behavior

The `joins(:source)` option gets removed from the final query causing it to fail (since the conditional references that table).

This was also reported in https://github.com/rails/rails/issues/22538 and closed by 
https://github.com/rails/rails/pull/29413 but that change didn't actually fix it. If you run the script from that issue you'll notice that it still fails.

### System configuration

This happens on Rails 5.1.6, 5.2.0, 5.2.1.rc1, and the master branch. Only tested with Ruby 2.5.1.

**Comment by twitnithegirl** [comment]
This is similar or related to this https://github.com/rails/rails/issues/33407

**Comment by imranmalik** [comment]
I have the same issue with Rails 5.0.7. Any fix or workaround? We are blocked on our product release because of this issue.

**Comment by twitnithegirl** [comment]
@imranmalik just to temper your expectations, 5.0.7 is no longer supported per the [Maintenance Policy](https://guides.rubyonrails.org/maintenance_policy.html) so once this does get fixed you'll have to upgrade your app to a new rails version or patch the issue yourself in your own implementation.

**Comment by imranmalik** [comment]
@twitnithegirl thank you for your note. We'll upgrade to Rails 5.2 gradually. Currently, we have to stick to 5.0.7 because of the protected_attributes_continued support.

Is there any patch or workaround available to make it work for 5.0.7?

**Comment by twitnithegirl** [comment]
Not yet @imranmalik  I don't think. I have some free time tomorrow and I'll see what I can figure out. I'll tag you if I come up with something.

**Comment by imranmalik** [comment]
@twitnithegirl thank you so much for your help.

**Comment by jeroenj** [comment]
> I have the same issue with Rails 5.0.7. Any fix or workaround?

A workaround in the above example is to change the `online_users` association to explicitly include the join too:

```ruby
has_many :online_users, -> { joins("INNER JOIN sources ON sources.id = comments.source_id") }, through: :online_comments, source: :user
```

**Comment by twitnithegirl** [comment]
thanks @jeroenj

**Comment by gregmolnar** [comment]
@ziaulrehman40 Are you sure you have the same issue? The problem described here is a missing JOIN clause, so if Rails generates the wrong JOIN clause for you, it is a different issue. Since you mentioned it is on 4.2, I am afraid there won't be much help, since that version is not supported anymore.

**Comment by ziaulrehman40** [comment]
Oh yeah, i am sorry, got here from this issue(which is my actual problem): https://github.com/rails/rails/issues/33407

Didn't realize this is a different issue. Sorry.

**Comment by gregmolnar** [comment]
@ziaulrehman40 that sounds like a totally unrelated issue.

**Comment by ziaulrehman40** [comment]
Well, not totally unrelated. Even first comment on this issue also points to that. Anyway, sorry for the confusion. Any help on the other issue would be great though.
(Basically i want to be able to OVERRIDE what join clause rails generates so i can typecast id from one table to match id column in other table)

**Comment by gregmolnar** [comment]
@ziaulrehman40 I would try to get around your issue with `unscope`.

**Comment by iczman** [comment]
I was scratching my head for hours only to find out this is due to a bug in rails... hope this will be fixed soon. @jeroenj's workaround works, thanks.

**Comment by toao** [comment]
I also stumbled over this issue today. It was a bit surprising. Maybe the documentation could be updated to explain this restriction.

**Comment by guitassinari** [comment]
I also faced this issue during the last weeks. It seems related to #29413 in some way. Maybe an edge case.

**Comment by jeroenj** [comment]
This can still be reproduced both on `master` and `5-2-stable`.

**Comment by jeroenj** [comment]
This can still be reproduced both on master and 5-2-stable.

**Comment by NoryDev** [comment]
This can still be reproduced both on master and 6-0-stable.

**Comment by gregmolnar** [comment]
This one still fails on 6.0.3.1.

**Comment by kaspth** [comment]
@gregmolnar that’s a security release, it won’t have other commits than what’s needed to fix the CVEs addressed in the release.

**Comment by gregmolnar** [comment]
@kaspth good point, so let me rephrase, it is still an issue on 6-0-stable.

**Comment by gregmolnar** [comment]
thank you @kamipo! you are a ⭐

**Comment by joemsak** [comment]
I still get it on 7.1.4

https://github.com/rails/rails/issues/44806#issuecomment-2334774904
